finetune/network_modules.py

```diff
@@ -350,3 +351,104 @@ def seq_lab_internal(hidden):
+                train=train, scale=False, mask=False
+            )
+            n = norm(attn_fn(hidden) + hidden, 'seq_label_residual')
+            flat_logits = tf.layers.dense(n, n_targets)
+            logits = tf.reshape(flat_logits, tf.concat([tf.shape(hidden)[:2], [n_targets]], 0))
+
+            association_head = tf.layers.dense(n, nx)
+            association_head = tf.reshape(association_head, tf.concat([tf.shape(hidden)[:2], [nx]], 0))
+
+            a = tf.expand_dims(association_head, 1)
+            b = tf.expand_dims(association_head, 2)
+
+            probs = tf.nn.softmax(logits)
+
+            features = tf.concat(
+                [
+                    a - b, a * b,
+                    tf.tile(a, [1, length, 1, 1]),
+                    tf.tile(b, [1, 1, length, 1]),
+                    tf.tile(tf.expand_dims(probs, 1),[1, length, 1, 1]),
+                    tf.tile(tf.expand_dims(probs, 2), [1, 1, length, 1])
+                ],
+                axis=-1
+            )
+            associations_flat = tf.layers.dense(
+                tf.reshape(features, shape=[-1, nx * 4 + 2 * n_targets])
+                , num_associations
+            )
+            associations = tf.reshape(associations_flat, [-1, length, length, num_associations])
+
+            return logits, associations_flat, associations
+
+        with tf.variable_scope('seq_lab_attn'):
+            if config.low_memory_mode and train:
+                seq_lab_internal = recompute_grad(seq_lab_internal, use_entire_scope=True)
+
+            logits, associations_flat, associations = seq_lab_internal(hidden)
+
+        log_likelihood = 0.0
+        association_loss = 0.0
+        class_weights = kwargs.get('class_weights')
+        if class_weights is not None:
+            logits = class_reweighting(class_weights)(logits)
+
+        transition_params = tf.get_variable("Transition_matrix", shape=[n_targets, n_targets])
+        if train and targets is not None:
+            log_likelihood, _ = crf_log_likelihood(
+                logits,
+                targets["labels"],
+                kwargs.get('max_length') * tf.ones(tf.shape(targets["labels"])[0]),
+                transition_params=transition_params
+            )
+            sequence_mask = tf.sequence_mask(pool_idx + 1, maxlen=length, dtype=tf.float32)
+            mask = tf.expand_dims(sequence_mask, 1) * tf.expand_dims(sequence_mask, 2)
+
+            association_loss = tf.losses.sparse_softmax_cross_entropy(
+                logits=associations_flat,
+                labels=tf.reshape(targets["associations"], shape=[-1]),
+                weights=tf.reshape(mask, shape=[-1])
+            )
+
+        association_loss = tf.Print(association_loss, [association_loss, -log_likelihood])
+
+        return {
+            'logits': {"sequence": logits, "association": associations},
+            'losses': -log_likelihood + 100 * association_loss,  # TODO: think about weighting.
```

Comment:
You already have this in a todo, but yeah definitely worth considering how we intend to make this weighting robust.

Reply:
Make this configurable